### ISSUE TITLE
bug: Sniper printer drops whitespace after comment if next element is type member with modified modifier list

### ISSUE THREAD

**Issue by slarse** [INRIA/spoon#3697]
I've encountered a problem where the sniper printer will not insert newlines correctly on modified or new elements. I've encountered this when inserting import statements, as well as when modifying fields. This is a big problem for the fields, because inline comments can cause a modified field to become commented out. For example, if we have this class:

```java
public class PrinterNewlines {
    // if a newline is not inserted on this line, the field below will be commented out
    public static int NON_FINAL_FIELD = 42;
}
```

And insert a modifier on the field (say `final`), this is the output:

```java
public class PrinterNewlines {
    // if a newline is not inserted on this line, the field below will be commented outpublic static final int NON_FINAL_FIELD = 42;
}
```

I'm working on figuring out why this happens.

**Comment by monperrus** [comment]
Thanks for the bug report, looking forward to the fix!

**Comment by slarse** [comment]
> Thanks for the bug report, looking forward to the fix!

So am I, this issue is messing with Sorald's patches :)

I've located the problem with the particular case I show in the opening post here, and it's a corner case of corner cases, in particular when there's a comment preceding a modified `CollectionSourceFragment` (e.g. modifiers). The problem of added import statements not getting newlines is related, but it's not exactly the same.

Overall, the logic for when to or when not to print newlines is a bit complicated, and gets really involved when elements are modified, added or deleted. I'm working on a fix, but it's slow going as I still don't fully understand the sniper printer.

**Comment by monperrus** [comment]
Cool. The expertise you're building while understanding the sniper 
architecture will be very valuable for both Sorald and Spoon.

**Comment by slarse** [comment]
I've determined that the import statement newlines is a separate issue, see #3698

**Comment by slarse** [comment]
Alright, after a ton of debugging, I've found that the problem here is as follows:

When there is a comment just preceding a type member (method, field, inner type), and the type member's modifiers are altered, things break down. The typical thing that happens is that the whitespace (newline + indentation) between the comment and the type member simply isn't printed. I've however also managed to get the comment to simply disappear.

There are also problems with modifiers on local variables. For example, adding the `final` keyword to a local variable that was preceded by a comment gave me this funky output:

```java
final
// comment here
int a = 2
```

I've found many fixes that have worked, but resulted in redundant whitespace elsewhere. I'll keep working on this, and will open a PR tomorrow with my test cases.
